### ISSUE TITLE
Requeue images if analyzer terminates gracefully

### ISSUE THREAD

**Issue by fischerman** [anchore/anchore-engine#296]
**Is this a request for help?**:

---
Yes

**Is this a BUG REPORT or a FEATURE REQUEST?** (choose one):
FEATURE REQUEST

**Version of Anchore Engine and Anchore CLI if applicable**:
0.5.1
 
**What happened**:
If an analyzer terminates gracefully (e.g. when scaling down) the image currently being analyzed remains in state `analyzing` until a timeout resets it (by default 10h).

**What did you expect to happen**:
The image should be requeued (before terminating) and picked up by another analyzer.

**Any relevant log output from /var/log/anchore**:

**What docker images are you using**:
Default images of helm chart `stable/anchore-engine`

**How to reproduce the issue**:
Fill up the queue and delete analyzer pods.

**Anything else we need to know**:

Especially in a Kubernetes setup it is common that pods are deleted (scaling, eviction, ...)

**Comment by fischerman** [comment]
Maybe related to that, the analyzer can unregister itself on termination (in a Kubernetes setup) because it's likely that the pod will be deleted and recreated under a new name (see #145).

**Comment by zhill** [comment]
There are 2 issues:
1. Clean shutdown. I any case where the container can exit cleanly it should de-register itself. This comes down to signal handling and may be deployment/runtime dependent
2. Updates to the k8s deploy to potentially use a stateful set instead of deployment for the analyzer workers so that names are stable. However, this seems like a bandaid that isn't necessary given that the analyzers are actually stateless (save for some caching). A statefulset seems like overkill here but could have advantages to preserve any caches in the scratch volume if that is persisted and re-used later.

**Comment by zhill** [comment]
Cleaner service shutdown is something that needs work. The challenge is the signal handling and how each platform sends signals to the containers to terminate. You're focused on k8s, correct?

**Comment by fischerman** [comment]
Yes, I think most platforms will issue a `SIGTERM` for termination, so this is not really k8s specific. I don't know whether the analyzer is already capable of properly stopping by returning the job back to the queue.

I guess the unregistration is more specific to k8s, because of unstable names.

**Comment by frittentheke** [comment]
Yes @zhill as @fischerman stated SIGTERM is the signal to handle here. This is not platform specific, but rather following the POSIX / Unix standard for this case of signalling an intended "termination" to a process. Be it a plain process, a process started / managed by a systemd unit or a fancy container -- A SIGTERM is shall receive to tell it to shut down :-)

**Comment by zhill** [comment]
Yes, SIGTERM is the right signal for sure, and I've done some initial prototyping for this issue already. Just things like timing from SIGTERM to SIGKILL if the process takes a big to respond, which we may given the need for network operations to update state of the service. I asked mostly for purposes of focused testing and other things like lifecycle hooks in k8s: https://kubernetes.io/docs/concepts/containers/container-lifecycle-hooks/#hook-details

**Comment by frittentheke** [comment]
The lifecycle hooks are nice, but to me using those only make sense if you cannot improve / implement this with the original piece of software and need "external" handlers. 

The default grade period for terminating pods is 30s (https://kubernetes.io/docs/concepts/workloads/pods/pod/#termination-of-pods). This should be plenty of time to stop the analyze job you are doing and to place the job back onto the queue, right?

**Comment by zhill** [comment]
> The lifecycle hooks are nice, but to me using those only make sense if you cannot improve / implement this with the original piece of software and need "external" handlers.
> 
> The default grace period for terminating pods is 30s (https://kubernetes.io/docs/concepts/workloads/pods/pod/#termination-of-pods). This should be plenty of time to stop the analyze job you are doing and to place the job back onto the queue, right?

Yes, others don't give that long, hence my question re: k8s or not. And its nice that its configurable. For now we'll try to keep all logic in the service so no hook is needed.

**Comment by StefanCenusa** [comment]
Hello! Is this enhancement already implemented or planned for any future release?
